extra/migrate_settings.rb

```diff
@@ -1,108 +1,228 @@
+      File.join(working_dir_path, migration_name)
+    end
+
+    def src_dir
+      File.join(migration_dir, "src")
+    end
+
+    def dst_dir
+      File.join(migration_dir, "dst")
+    end
+
+    def underscore(src)
+      src = src.gsub(/::/, '/')
+      src = src.gsub(/([A-Z]+)([A-Z][a-z])/,'\1_\2')
+      src = src.gsub(/([a-z\d])([A-Z])/,'\1_\2')
+      src = src.tr("-", "_")
+      src.downcase
+    end
   end
 
-  # Rename the port to whichever is correct
-  if output[:settings].keys.include?(:port)
-    if output[:settings].keys.include?(:ssl_certificate)
-      output[:settings][:https_port] = output[:settings].delete(:port)
-    else
-      output[:settings][:http_port] = output[:settings].delete(:port)
+  class Migrator
+    attr_reader :working_dir_path, :migrations_dir_path, :config_src_path, :modules_config_src_path, :result_dir_path
+
+    def initialize(working_dir_path, migrations_dir_path, config_src_path, modules_config_src_path, migrations)
+      @working_dir_path = working_dir_path
+      @migrations_dir_path = migrations_dir_path
+      @config_src_path = config_src_path
+      @modules_config_src_path = modules_config_src_path
+      @migrations = migrations
+      @result_dir_path = File.join(working_dir_path, 'result')
+    end
+
+    def load_migrations(migrations_dir_path)
+      migration_files = Dir.glob(File.join(migrations_dir_path, "*.rb")).sort
+      ::Proxy::Migration.inject_migrations_instance(@migrations)
+      migration_files.each { |file| require file }
+    end
+
+    def migrate
+      raise "Migrations dir '#{migrations_dir_path}' doesn't exist" unless Dir.exist?(migrations_dir_path)
+      raise "Settings file '#{config_src_path}' doesn't exist" unless File.exist?(config_src_path)
+
+      used_values = %{using:
+  config file: #{config_src_path},
+  modules config dir: #{modules_config_src_path},
+  working dir: #{working_dir_path},
+  migrations dir: #{migrations_dir_path},
+  migrations state path: #{@migrations.migration_state_file_path}
+      }
+      puts used_values
+
+      FileUtils.mkdir_p(result_dir_path)
+      load_migrations(migrations_dir_path)
+
+      puts "Running migrations..."
+      migrations = @migrations.new_migrations
+      migrations.each do |migration|
+        m = migration.new(working_dir_path)
+        puts "#{m.migration_name}"
+
+        create_migration_dirs(m.src_dir, m.dst_dir)
+        if migration == migrations.first
+          copy_original_configuration(m.src_dir)
+        else
+          copy_previous_migration_results(@migrations.executed_migrations.last.dst_dir, m.src_dir)
+        end
+
+        m.migrate
+        @migrations.executed_migrations << m
+      end
+
+      if @migrations.executed_migrations.empty?
+        copy_original_configuration(result_dir_path)
+      else
+        copy_to_results_dir(@migrations.executed_migrations.last.dst_dir, result_dir_path)
+      end
+    ensure
+      @migrations.persist_migrations_state(result_dir_path)
```

Comment:
This can hide exceptions when they're thrown at the start of this method, as result_dir_path doesn't exist until the mkdir_p call on line 114:

<pre>
# ruby /usr/share/foreman-proxy/extra/migrate_settings.rb 
/usr/share/foreman-proxy/extra/migrate_settings.rb:34:in `initialize': No such file or directory - /usr/share/foreman-proxy/tmp/result/migration_state (Errno::ENOENT)
        from /usr/share/foreman-proxy/extra/migrate_settings.rb:34:in `open'
        from /usr/share/foreman-proxy/extra/migrate_settings.rb:34:in `persist_migrations_state'
        from /usr/share/foreman-proxy/extra/migrate_settings.rb:140:in `migrate'
        from /usr/share/foreman-proxy/extra/migrate_settings.rb:224
</pre>